src/mlpack/core/tree/rectangle_tree/r_plus_tree_split_impl.hpp

```diff
@@ -0,0 +1,342 @@
+      child->Parent() = treeOne;
+    }
+    else if (policy == SplitPolicyType::AssignToSecondTree)
+    {
+      InsertNodeIntoTree(treeTwo, child);
+      child->Parent() = treeTwo;
+    }
+    else
+    {
+      // The child should be split (i.e. the partition divides its bound).
+      TreeType* childOne = new TreeType(treeOne);
+      TreeType* childTwo = new TreeType(treeTwo);
+      treeOne->MinLeafSize() = 0;
+      treeOne->MinNumChildren() = 0;
+      treeTwo->MinLeafSize() = 0;
+      treeTwo->MinNumChildren() = 0;
+
+      // Propagate the split downward.
+      if (child->IsLeaf())
+        SplitLeafNodeAlongPartition(child, childOne, childTwo, cutAxis, cut);
+      else
+        SplitNonLeafNodeAlongPartition(child, childOne, childTwo, cutAxis, cut);
+
+      InsertNodeIntoTree(treeOne, childOne);
+      InsertNodeIntoTree(treeTwo, childTwo);
+
+      child->SoftDelete();
+    }
+  }
+
+  assert(treeOne->NumChildren() + treeTwo->NumChildren() != 0);
+
+  // Add a fake subtree if one of the subtrees is empty.
+  if (treeOne->NumChildren() == 0)
+    AddFakeNodes(treeTwo, treeOne);
+  else if (treeTwo->NumChildren() == 0)
+    AddFakeNodes(treeOne, treeTwo);
+
+  assert(treeOne->NumChildren() <= treeOne->MaxNumChildren());
+  assert(treeTwo->NumChildren() <= treeTwo->MaxNumChildren());
+}
+
+template<typename SplitPolicyType,
+         template<typename> class SweepType>
+template<typename TreeType>
+void RPlusTreeSplit<SplitPolicyType, SweepType>::
+AddFakeNodes(const TreeType* tree, TreeType* emptyTree)
+{
+  size_t numDescendantNodes = tree->TreeDepth() - 1;
+
+  TreeType* node = emptyTree;
+  for (size_t i = 0; i < numDescendantNodes; i++)
+  {
+    TreeType* child = new TreeType(node);
+    node->children[node->NumChildren()++] = child;
+
+    node = child;
+  }
+}
+
+template<typename SplitPolicyType,
+         template<typename> class SweepType>
+template<typename TreeType>
+bool RPlusTreeSplit<SplitPolicyType, SweepType>::
+PartitionNode(const TreeType* node, size_t& minCutAxis,
+    typename TreeType::ElemType& minCut)
+{
+  if ((node->NumChildren() <= fillFactor && !node->IsLeaf()) ||
+      (node->Count() <= fillFactor && node->IsLeaf()))
+    return false; // No partition required.
+
+  // Define the type of the sweep cost.
+  typedef typename
+      SweepType<SplitPolicyType>::template SweepCost<TreeType>::type
+      SweepCostType;
+
+  SweepCostType minCost = std::numeric_limits<SweepCostType>::max();
+  minCutAxis = node->Bound().Dim();
+
+  // Find the sweep with a minimal cost.
+  for (size_t k = 0; k < node->Bound().Dim(); k++)
+  {
+    typename TreeType::ElemType cut;
+    SweepCostType cost;
+
+    if (node->IsLeaf())
+      cost = SweepType<SplitPolicyType>::SweepLeafNode(k, node, cut);
+    else
+      cost = SweepType<SplitPolicyType>::SweepNonLeafNode(k, node, cut);
+    
+
+    if (cost < minCost)
+    {
+      minCost = cost;
+      minCutAxis = k;
+      minCut = cut;      
+    }
+  }
+  return true;
```

Comment:
But I think I can produce that behavior for the R++ tree if all the points are 0 in all dimensions. Then `SplitLeafNode()` in every dimension will be `SIZE_MAX`. I think a similar thing will happen for non-leaf nodes or for the R+ tree.

I guess in that type of situation, when we cannot split the node, we have to either raise the maximum number of points in each node, or throw an exception. I'd probably lean towards the latter, with a line of output to `Log::Warn` in order to warn the user.